packages/grafana-ui/src/components/DateTimePickers/TimeRangeContext.tsx

```diff
@@ -0,0 +1,104 @@
+  removePicker(): void;
+};
+
+export type TimeRangeContextHookValue = {
+  // If the time range is synced, this is the value that all pickers should show.
+  syncedValue?: TimeRange;
+
+  // If the time range is synced across all visible pickers.
+  synced: boolean;
+
+  // If it is possible to sync the time range. This is based just on the number of pickers that are visible so if
+  // there is only one, the sync button can be hidden.
+  syncPossible: boolean;
+
+  // Action passed to the picker to interact with the sync state.
+  // Sync the time range across all pickers with the provided value. Can be also used just to update a value when
+  // already synced.
+  sync(value: TimeRange): void;
+  unSync(): void;
+};
+
+const TimeRangeContext = React.createContext<TimeRangeContextValue | undefined>(undefined);
+
+export function TimeRangeProvider({ children }: { children: React.ReactNode }) {
+  // We simply keep the count of the pickers visible by letting them call the addPicker and removePicker functions.
+  const [pickersCount, setPickersCount] = useState(0);
+  const [synced, setSynced] = useState(false);
+  const [syncedValue, setSyncedValue] = useState<TimeRange>();
+
+  const sync = useCallback((value: TimeRange) => {
+    setSynced(true);
+    setSyncedValue(value);
+  }, []);
+
+  const unSync = useCallback(() => {
+    setSynced(false);
+    setSyncedValue(undefined);
+  }, []);
+
+  const contextVal = useMemo(() => {
+    return {
+      sync,
+      unSync,
+      addPicker: () => setPickersCount((val) => val + 1),
+      removePicker: () => setPickersCount((val) => val - 1),
+      syncPossible: pickersCount > 1,
+      synced,
+      syncedValue,
+    };
+  }, [pickersCount, sync, unSync, synced, syncedValue]);
+
+  return <TimeRangeContext.Provider value={contextVal}>{children}</TimeRangeContext.Provider>;
+}
+
+export function useTimeRangeContext(initialSyncValue?: TimeRange): TimeRangeContextHookValue | undefined {
+  const context = React.useContext(TimeRangeContext);
+
+  // Automatically add and remove the picker when the component mounts and unmounts or if context changes (but that
+  // should not happen). We ignore the initialSyncValue to make this value really just an initial value and isn't a
+  // prop by which you could control the picker.
+  useEffect(() => {
+    // We want the pickers to still function even if they are not used in a context. Not sure, this will be a common
+    // usecase, but it does not seem like it will cost us anything.
+    if (context) {
+      context.addPicker();
+      if (initialSyncValue) {
+        context.sync(initialSyncValue);
+      }
+      return () => {
+        context.removePicker();
+      };
+    }
+    return () => {};
+    // We want to do this only on mount and unmount.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  return useMemo(() => {
+    if (!context) {
+      return context;
+    }
```

Comment:
Not sure I understand this part, does it effectively return undefined when there's no context?

Reply:
Yes, this looks a bit weird I guess but the only reason to do this instead of throwing an error if the provider isn't there is so the time picker will work without it if used somewhere else than in Grafana main render context. Not sure if that is super useful (maybe some testing envs, or if somebody uses grafana/ui as a standalone UI lib) but at the same time I don't think I would want to make the provider into such strong dependency.

Reply:
Added, well duplicated a comment here to explain it.

Reply:
Agree that throwing an error would be too disruptive, but maybe we could add some warning for this case, to help with debugging?

Reply:
But is it usable/going to be used outside the provider?

Reply:
So right now if you use the TimePicker outside of a provider it will just not have the sync button and you won't be able to sync with others but everything will work as before. Only one test for the picker include the provider (to see if we show the sync button in that case) others should work as before

Reply:
Ah ok, got it. Let's keep it like this then.